cups/man/lpoptions.man
.\"
.\" "$Id: lpoptions.man,v 1.1.1.9 2004/06/05 02:42:32 jlovell Exp $"
.\"
.\"   lpoptions man page for the Common UNIX Printing System (CUPS).
.\"
.\"   Copyright 1997-2003 by Easy Software Products.
.\"
.\"   These coded instructions, statements, and computer programs are the
.\"   property of Easy Software Products and are protected by Federal
.\"   copyright law.  Distribution and use rights are outlined in the file
.\"   "LICENSE.txt" which should have been included with this file.  If this
.\"   file is missing or damaged please contact Easy Software Products
.\"   at:
.\"
.\"       Attn: CUPS Licensing Information
.\"       Easy Software Products
.\"       44141 Airport View Drive, Suite 204
.\"       Hollywood, Maryland 20636-3111 USA
.\"
.\"       Voice: (301) 373-9603
.\"       EMail: cups-info@cups.org
.\"         WWW: http://www.cups.org
.\"
.TH lpoptions 1 "Common UNIX Printing System" "21 October 2002" "Easy Software Products"
.SH NAME
lpoptions \- display or set printer options and defaults
.SH SYNOPSIS
.B lpoptions
[ -h
.I server
] -d
.I dest[/instance]
[ -o
.I option=value
] ... [ -o
.I option=value
]
.br
.B lpoptions
[ -h
.I server
] [ -E ] [ -p
.I dest[/instance]
] -l
.br
.B lpoptions
[ -h
.I server
] [ -E ] [ -o
.I option=value
] ... [ -o
.I option=value
] [ -p
.I dest[/instance]
] -r
.I option
.br
.B lpoptions
[ -h
.I server
] [ -E ] -x
.I dest[/instance]
.SH DESCRIPTION
\fBlpoptions\fR displays or sets printer options and defaults.
\fBlpoptions\fR shows the default printer options when run with no
arguments. Other options include:
.TP 5
\-E
.br
Enables encryption when communicating with the CUPS server.
.TP 5
\-d \fIdest[/instance]\fR
.br
Sets the default printer to \fIdest\fR. If \fIinstance\fR is supplied then
that particular instance is used. This option overrides the system default
printer for the current user.
.TP 5
\-h \fIserver\fR
.br
Specifies the CUPS server to communicate with.
.TP 5
\-l
.br
Lists the printer specific options and their current settings.
.TP 5
\-o \fIoption=value\fR
.br
Specifies a new option for the named destination.
.TP 5
\-p \fIdest[/instance]\fR
.br
Sets the destination and instance, if specified, for any options that follow.
If the named instance does not exist then it is created.
.TP 5
\-r \fIoption\fR
.br
Removes the specified option for the named destination.
.TP 5
\-x \fIdest[/instance]\fR
.br
Removes the options for the named destination and instance, if specified.
If the named instance does not exist then this does nothing.
.LP
If no options are specified using the \fI-o\fR option then the current
options for the named printer are reported on the standard output.
.LP
Options set with the \fBlpoptions\fR command are used by the \fBlp(1)\fR
and \fBlpr(1)\fR commands when submitting jobs.
.SH ROOT ACCOUNT OPTIONS
When run by the root user, \fBlpoptions\fR gets and sets default
options and instances for \fIall users\fR in the
/etc/cups/lpoptions file.
.SH COMPATIBILITY
The \fBlpoptions\fR command is unique to CUPS.
.SH FILES
~/.lpoptions - user defaults and instances created by non-root users.
.br
/etc/cups/lpoptions - system-wide defaults and instances created by the root
user.
.SH SEE ALSO
cancel(1), lp(1),
CUPS Software Users Manual,
http://localhost:631/documentation.html
.SH COPYRIGHT
Copyright 1993-2003 by Easy Software Products, All Rights Reserved.
.\"
.\" End of "$Id: lpoptions.man,v 1.1.1.9 2004/06/05 02:42:32 jlovell Exp $".
.\"


QuickTimeStreamingServer/ReleaseNotes.txt

                                 RELEASE NOTES
                             Darwin Streaming Server 

These release notes are for the current development version of Darwin
Streaming Server.

------------------------------------------------------------------------
Date: 8/7/2001
Version: 3.0.1+development [v349]
cvs tag: qtssServer-349

*   Updated build platforms: Solaris 8, Redhat Linux 7.1, FreeBSD 4.3, 
    Mac OS X 10.1
*   Retransmit/overbuffering (Skip Protection) algorithms tuned
    substantially
*   "reliable_udp" attribute defaults to "enabled" 
*   Various WebAdmin fixes for Windows platform

------------------------------------------------------------------------
Date: 8/9/2001
Version: 3.0.1+development [v350]
cvs tag: qtssServer-350

*   Minor bug fixes.

------------------------------------------------------------------------
Date: 8/14/2001
Version: 3.0.1+development [v351]
cvs tag: qtssServer-351

*   QTSSFileModule - enabled caching based on preferences (shared and
    private).
*   Add buffer delay to on demand movies.
*   QTFile - code cleanup.
*   Changed some paths for the installer.
*   Added support for MP3Broadcaster on Windows.

------------------------------------------------------------------------
Date: 8/16/2001
Version: 3.0.1+development [v352]
cvs tag: qtssServer-352

*   Added MP3StreamingModule to all builds.
*   Fixed gdb script to execute QTSS from build dir on MacOSX.

------------------------------------------------------------------------
Date: 8/19/2001
Version: 3.0.1+development [v353]
cvs tag: qtssServer-353

*   Added:
        add_seconds_to_client_buffer_delay -- Increases client buffer
                                delay before playing movie.
        record_movie_file_sdp -- Records a .sdp file for a .mov file.
                                Appears as moviename.mov.sdp
        enable_movie_file_sdp -- Plays a .mov.sdp file for a .mov movie.
*   Fixed server to play .mov.sdp files.
*   QTSSFileModule enables both private and shared movie cache buffers.

------------------------------------------------------------------------
Date: 8/23/2001
Version: 3.0.1+development [v354]
cvs tag: qtssServer-354

*   Added prefs for MP3StreamingModule.
*   Default Reliable UDP dir is / on UNIX and \ on NT
*   Use Enable Reliable UDP pref to disable Reliable UDP
*   Included new Web Admin
*   Changed relay preferences to XML format.
        Old relay config files will not work.

------------------------------------------------------------------------
Date: 08/28/2001
Version: 3.0.1+development [v355]
cvs tag: qtssServer-355

*	Added preference:
	 RTSP_error_message to enable debugging error strings returned in 
	 the content body of RTSP error responses.
	
*	Changed RTSP code to:
	1) require a Cseq field for OPTIONS requests
	2) allow missing Range headers in a PLAY request (defaults to entire movie)
	3) add content length header to Error responses containing error messages
	4) add SDP o= and t= fields to on-demand movie Describe responses

*      Changed Relay to relay incoming ANNOUNCEd broadcasts
       1) All ANNOUNCEd broadcasts, or
       2) All broadcasts from an IP address
       3) A broadcast URL from an IP address

       (a) Can still configure the relay to relay UDP streams
       (b) Can pull an SDP or movie from source and relay it

*      Changed Relay to ANNOUNCE the SDPs to its destinations
       1) for ANNOUNCED sources
       2) for sources that are pulled from the source server

*      Changed Relay config file to be in XML format
       1) relayconfig.xml-Sample will be installed instead of old streamingrelay.conf

*	Added General Settings and Change Password to UI
	1) Require existing password to change to new
	2) Require the user to enter the same password twice
	3) Change the auth cookie so the user won't have to re-authenticate

-------------------------------------------------------------------------------------------------
Date: 8/31/2001
Version: 3.0.1+development [v356]
cvs tag: qtssServer-356

*	 Admin Module change
	 1) Added an admin module pref called "AdministratorGroup" with
	 a default value of "admin"
	 2) Changed Admin Module to allow all users that belong to the
	 AdministratorGroup, instead of allowing only one user called
	 "streamingadmin".
	 3) A default qtgroups gets installed from now. The default qtusers
	 file has user 'streamingadmin' with password 'default'. The default
	 qtgroups file has the group 'admin' with the user 'streamingadmin'.
	 4) The login name for the web admin can be changed by adding the user
	 to the qtusers file and adding the user to the 'admin' group in the
	 qtgroups file.

*	Added "RTSP_debug_printfs" server pref to show the incoming and outgoing RTSP.
	Fixed Session header response to remove ";timeout=" field if the value is not defined.
	Fixed .mov.sdp Describe response to include the Last-Modified and Cache-Control headers.
	Fixed RTSP Session header parser to also stop on a ';'
	Changes shared file buffer size to default to maximum size if the bit-rate is 0 and no fixed 
	buffer size is defined.

-------------------------------------------------------------------------------------------------
Date: 9/04/2001
Version: 3.0.1+development [v357]
cvs tag: qtssServer-357

*	more implementation code for Web admin added.
*	New XML parser for reading prefs.
*	full implementation of MP3 streaming module is included.

-------------------------------------------------------------------------------------------------
Date: 9/06/2001
Version: 3.0.1+development [v359]
cvs tag: qtssServer-359

*	Relay that pulls source from the same server can be setup now
	[rtsp_source SOURCE with own IP address as the source_addr]
   
-------------------------------------------------------------------------------------------------
Date: 9/10/2001
Version: 3.0.1+development [v361]
cvs tag: qtssServer-361

*	UI: play and stop buttons in playlist screen now start/stop playlists immediately

cups/systemv/lpoptions.c
/*
 * "$Id: lpoptions.c,v 1.1.1.13 2004/06/05 02:42:35 jlovell Exp $"
 *
 *   Printer option program for the Common UNIX Printing System (CUPS).
 *
 *   Copyright 1997-2004 by Easy Software Products.
 *
 *   These coded instructions, statements, and computer programs are the
 *   property of Easy Software Products and are protected by Federal
 *   copyright law.  Distribution and use rights are outlined in the file
 *   "LICENSE.txt" which should have been included with this file.  If this
 *   file is missing or damaged please contact Easy Software Products
 *   at:
 *
 *       Attn: CUPS Licensing Information
 *       Easy Software Products
 *       44141 Airport View Drive, Suite 204
 *       Hollywood, Maryland 20636-3111 USA
 *
 *       Voice: (301) 373-9603
 *       EMail: cups-info@cups.org
 *         WWW: http://www.cups.org
 *
 * Contents:
 *
 *   main()         - Main entry.
 *   list_group()   - List printer-specific options from the PPD group.
 *   list_options() - List printer-specific options from the PPD file.
 *   usage()        - Show program usage and exit.
 */

/*
 * Include necessary headers...
 */

#include <cups/cups.h>
#include <cups/string.h>
#include <stdlib.h>


/*
 * Local functions...
 */

void	list_group(ppd_group_t *group);
void	list_options(cups_dest_t *dest);
void	usage(void);


/*
 * 'main()' - Main entry.
 */

int				/* O - Exit status */
main(int  argc,			/* I - Number of command-line arguments */
     char *argv[])		/* I - Command-line arguments */
{
  int		i, j;		/* Looping vars */
  int		changes;	/* Did we make changes? */
  int		num_options;	/* Number of options */
  cups_option_t	*options;	/* Options */
  int		num_dests;	/* Number of destinations */
  cups_dest_t	*dests;		/* Destinations */
  cups_dest_t	*dest;		/* Current destination */
  char		*printer,	/* Printer name */
		*instance,	/* Instance name */ 
 		*option;	/* Current option */


 /*
  * Loop through the command-line arguments...
  */

  dest        = NULL;
  num_dests   = 0;
  dests       = NULL;
  num_options = 0;
  options     = NULL;
  changes     = 0;

  for (i = 1; i < argc; i ++)
    if (argv[i][0] == '-')
    {
      switch (argv[i][1])
      {
        case 'd' : /* -d printer */
	    if (argv[i][2])
	      printer = argv[i] + 2;
	    else
	    {
	      i ++;
	      if (i >= argc)
	        usage();

	      printer = argv[i];
	    }

            if ((instance = strrchr(printer, '/')) != NULL)
	      *instance++ = '\0';

	    if (num_dests == 0)
	      num_dests = cupsGetDests(&dests);

            if ((dest = cupsGetDest(printer, instance, num_dests, dests)) == NULL)
	    {
	      fputs("lpoptions: Unknown printer or class!\n", stderr);
	      return (1);
	    }

	   /*
	    * Set the default destination...
	    */

	    for (j = 0; j < num_dests; j ++)
	      dests[j].is_default = 0;

	    dest->is_default = 1;

	    cupsSetDests(num_dests, dests);

	    for (j = 0; j < dest->num_options; j ++)
	      if (cupsGetOption(dest->options[j].name, num_options, options) == NULL)
		num_options = cupsAddOption(dest->options[j].name,
	                                    dest->options[j].value,
	                                    num_options, &options);
	    break;

	case 'h' : /* -h server */
	    if (argv[i][2])
	      cupsSetServer(argv[i] + 2);
	    else
	    {
	      i ++;
	      if (i >= argc)
	        usage();

	      cupsSetServer(argv[i]);
	    }
	    break;

        case 'E' : /* Encrypt connection */
	    cupsSetEncryption(HTTP_ENCRYPT_REQUIRED);
	    break;

	case 'l' : /* -l (list options) */
            if (dest == NULL)
	    {
	      if (num_dests == 0)
		num_dests = cupsGetDests(&dests);

	      if ((dest = cupsGetDest(NULL, NULL, num_dests, dests)) == NULL)
	        dest = dests;
	    }

            if (dest == NULL)
	      fputs("lpoptions: No printers!?!\n", stderr);
	    else
	      list_options(dest);

            changes = -1;
	    break;

	case 'o' : /* -o option[=value] */
	    if (argv[i][2])
	      num_options = cupsParseOptions(argv[i] + 2, num_options, &options);
	    else
	    {
	      i ++;
	      if (i >= argc)
	        usage();

	      num_options = cupsParseOptions(argv[i], num_options, &options);
	    }

	    changes = 1;
	    break;

	case 'p' : /* -p printer */
	    if (argv[i][2])
	      printer = argv[i] + 2;
	    else
	    {
	      i ++;
	      if (i >= argc)
	        usage();

	      printer = argv[i];
	    }

            if ((instance = strrchr(printer, '/')) != NULL)
	      *instance++ = '\0';

	    if (num_dests == 0)
	      num_dests = cupsGetDests(&dests);

            if ((dest = cupsGetDest(printer, instance, num_dests, dests)) == NULL)
	    {
	      num_dests = cupsAddDest(printer, instance, num_dests, &dests);
	      dest      = cupsGetDest(printer, instance, num_dests, dests);

              if (dest == NULL)
	      {
	        perror("lpoptions: Unable to add printer or instance");
		return (1);
	      }
	    }

	    for (j = 0; j < dest->num_options; j ++)
	      if (cupsGetOption(dest->options[j].name, num_options, options) == NULL)
		num_options = cupsAddOption(dest->options[j].name,
	                                    dest->options[j].value,
	                                    num_options, &options);
	    break;

	case 'r' : /* -r option (remove) */
	    if (argv[i][2])
	      option = argv[i] + 2;
	    else
	    {
	      i ++;
	      if (i >= argc)
	        usage();

	      option = argv[i];
	    }

            for (j = 0; j < num_options; j ++)
	      if (strcasecmp(options[j].name, option) == 0)
	      {
	       /*
	        * Remove this option...
		*/

	        num_options --;

		if (j < num_options)
		  memcpy(options + j, options + j + 1,
		         sizeof(cups_option_t) * (num_options - j));
		break;
              }

	    changes = 1;
	    break;

        case 'x' : /* -x printer */
	    if (argv[i][2])
	      printer = argv[i] + 2;
	    else
	    {
	      i ++;
	      if (i >= argc)
	        usage();

	      printer = argv[i];
	    }

            if ((instance = strrchr(printer, '/')) != NULL)
	      *instance++ = '\0';

	    if (num_dests == 0)
	      num_dests = cupsGetDests(&dests);

            if ((dest = cupsGetDest(printer, instance, num_dests, dests)) != NULL)
	    {
              cupsFreeOptions(dest->num_options, dest->options);

             /*
	      * If we are "deleting" the default printer, then just set the
	      * number of options to 0; if it is also the system default
	      * then cupsSetDests() will remove it for us...
	      */

	      if (dest->is_default)
	      {
		dest->num_options = 0;
		dest->options     = NULL;
	      }
	      else
	      {
		num_dests --;

		j = dest - dests;
		if (j < num_dests)
		  memcpy(dest, dest + 1, (num_dests - j) * sizeof(cups_dest_t));
	      }
	    }

	    cupsSetDests(num_dests, dests);
	    dest    = NULL;
	    changes = -1;
	    break;

	default :
	    usage();
      }
    }
    else
      usage();

  if (num_dests == 0)
    num_dests = cupsGetDests(&dests);

  if (dest == NULL)
  {
    if ((dest = cupsGetDest(NULL, NULL, num_dests, dests)) != NULL)
    {
      for (j = 0; j < dest->num_options; j ++)
	if (cupsGetOption(dest->options[j].name, num_options, options) == NULL)
	  num_options = cupsAddOption(dest->options[j].name,
	                              dest->options[j].value,
	                              num_options, &options);
    }
  }

  if (dest == NULL)
    return (0);

  if (changes > 0)
  {
   /*
    * Set printer options...
    */

    cupsFreeOptions(dest->num_options, dest->options);

    dest->num_options = num_options;
    dest->options     = options;

    cupsSetDests(num_dests, dests);
  }
  else if (changes == 0)
  {
    num_options = dest->num_options;
    options     = dest->options;

    for (i = 0; i < num_options; i ++)
    {
      if (i)
        putchar(' ');

      if (!options[i].value[0])
        printf("%s", options[i].name);
      else if (strchr(options[i].value, ' ') != NULL ||
               strchr(options[i].value, '\t') != NULL)
	printf("%s=\'%s\'", options[i].name, options[i].value);
      else
	printf("%s=%s", options[i].name, options[i].value);
    }

    putchar('\n');
  }

  return (0);
}

/*
 * 'list_group()' - List printer-specific options from the PPD group.
 */

void
list_group(ppd_group_t *group)	/* I - Group to show */
{
  int		i, j;		/* Looping vars */
  ppd_option_t	*option;	/* Current option */
  ppd_choice_t	*choice;	/* Current choice */
  ppd_group_t	*subgroup;	/* Current subgroup */


  for (i = group->num_options, option = group->options; i > 0; i --, option ++)
  {
    printf("%s/%s:", option->keyword, option->text);

    for (j = option->num_choices, choice = option->choices; j > 0; j --, choice ++)
      if (choice->marked)
        printf(" *%s", choice->choice);
      else
        printf(" %s", choice->choice);

    putchar('\n');
  }

  for (i = group->num_subgroups, subgroup = group->subgroups; i > 0; i --, subgroup ++)
    list_group(subgroup);
}


/*
 * 'list_options()' - List printer-specific options from the PPD file.
 */

void
list_options(cups_dest_t *dest)	/* I - Destination to list */
{
  int		i;		/* Looping var */
  const char	*filename;	/* PPD filename */
  ppd_file_t	*ppd;		/* PPD data */
  ppd_group_t	*group;		/* Current group */


  if ((filename = cupsGetPPD(dest->name)) == NULL)
  {
    fprintf(stderr, "lpoptions: Destination %s has no PPD file!\n", dest->name);
    return;
  }

  if ((ppd = ppdOpenFile(filename)) == NULL)
  {
    unlink(filename);
    fprintf(stderr, "lpoptions: Unable to open PPD file for %s!\n", dest->name);
    return;
  }

  ppdMarkDefaults(ppd);
  cupsMarkOptions(ppd, dest->num_options, dest->options);

  for (i = ppd->num_groups, group = ppd->groups; i > 0; i --, group ++)
    list_group(group);

  ppdClose(ppd);
  unlink(filename);
}


/*
 * 'usage()' - Show program usage and exit.
 */

void
usage(void)
{
  puts("Usage: lpoptions [-h server] [-E] -d printer");
  puts("       lpoptions [-h server] [-E] [-p printer] -l");
  puts("       lpoptions [-h server] [-E] -p printer -o option[=value] ...");
  puts("       lpoptions [-h server] [-E] -x printer");

  exit(1);
}


/*
 * End of "$Id: lpoptions.c,v 1.1.1.13 2004/06/05 02:42:35 jlovell Exp $".
 */
